sycl/doc/extensions/experimental/sycl_ext_oneapi_spec_constant_length_alloca.asciidoc

```diff
@@ -0,0 +1,195 @@
+change incompatibly in future versions of {dpcpp} without prior notice.
+*Shipping software products should not rely on APIs defined in this
+specification.*
+
+== Overview
+
+C++ arrays and `std::array` can be used in SYCL code to represent fixed-size
+sequences of objects. However, these constructs have a significant restriction:
+the number of elements must be known at compile time. In host-code context,
+users can make use of dynamic memory allocations, e.g., `std::vector`, but this
+is not the case in SYCL device code.
+
+SYCL specialization constants (SYCL Section 4.9.5.) can be used to represent
+constants whose values can be set dynamically during the execution of a SYCL
+application, but that will not change when a SYCL kernel function is
+invoked. This way, specialization constants could be used to implement SYCL
+private arrays whose size is given during the execution of the SYCL
+application. There is no possible way of implementing this using `std::array`,
+as the size of such container must be known at compile time, so we propose to
+define a new `private_alloca` function whose size is specified using SYCL
+specialization constants.
+
+[NOTE]
+====
+This extension only supports SPIR-V backends for now, as it relies on
+SPIR-V-specific capabilities,
+such as specialization constants.
+====
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_SPEC_CONSTANT_LENGTH_ALLOCA` to one of the values defined
+in the table below. Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== The `private_alloca` function
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+template <typename ElementType, auto &SpecName,
+          access::decorated DecorateAddress>
+private_ptr<ElementType, DecorateAddress>
+private_alloca(kernel_handler &h);
+} // namespace sycl::ext::oneapi::experimental
+----
+
+This extension adds a new `private_alloca` function that can be used to allocate
+a private memory region with capacity for
+`h.get_specialization_constant<SpecName>()` elements of type `ElementType`. The
+returned pointer will be aligned for `ElementType`.
+
+The underlying memory region is automatically freed when the caller to
+`private_alloca` is returned.
+
+In case of private memory exhaustion, the underlying backend must report an
+error in the same fashion as if the allocation size were static.
+
+If this function is called from host context or an unsupported backend, an
```

Comment:
Just checking -- this means that the function needs to be defined for all backends, even if it doesn't work.  Could/should this functionality be hidden behind an aspect instead?

Reply:
I defined an aspect for that. Thanks!